storage/index/block.go

```diff
@@ -111,86 +108,74 @@ func (b *block) EndTime() time.Time {
 	return b.endTime
 }
 
-func (b *block) updateWriteBatchWithLock(inserts []WriteBatchEntry) {
-	for _, insert := range inserts {
-		b.writeBatch.Docs = append(b.writeBatch.Docs, insert.Document)
-	}
-}
-
-func (b *block) resetWriteBatchWithLock() {
-	var emptyDoc doc.Document
-	for i := range b.writeBatch.Docs {
-		b.writeBatch.Docs[i] = emptyDoc
-	}
-	b.writeBatch.Docs = b.writeBatch.Docs[:0]
-}
-
-func (b *block) WriteBatch(inserts []WriteBatchEntry) (WriteBatchResult, error) {
-	b.Lock()
-	defer func() {
-		b.resetWriteBatchWithLock()
-		b.Unlock()
-	}()
+func (b *block) WriteBatch(inserts WriteBatchEntryByBlockStartAndID) (WriteBatchResult, error) {
+	b.RLock()
+	defer b.RUnlock()
 
 	if b.state != blockStateOpen {
 		// NB: releasing all references to inserts
-		for _, insert := range inserts {
-			insert.OnIndexSeries.OnIndexFinalize()
-		}
+		WriteBatchEntriesFinalizer(inserts).Finalize()
```

Comment:
when you revert to the earlier version, would be good to retain this utility type - `WriteBatchEntriesFinalizer`